README.txt
If you want to replace the t3blog commenting interface with mm_forum, insert the
static TypoScript template that contains the word "REPLACE"


Hint:
If you use t3blog commenting function mm_forum_blog creates only copies of the comments. It is recommended not to make the topic / forum writeable for everybody, because mm_forum doesn't create t3_blog comments if mm_forum posts are created (TODO).


Known "bugs" / Missing features:
- Works only with "page.10.widget.blogList.approved = 1" as the backend module creates a link to "tce_db.php" (with a command like copy, move,delete for records)
    See: http://typo3.org/documentation/document-library/core-documentation/doc_core_api/4.3.0/view/3/2/ -> Keyword: issueCommand
    See also: http://typo3.org/documentation/document-library/core-documentation/doc_core_api/4.3.0/view/3/3/#id2507501   
- You cannot comment comments as it is not yet supported by mm_forum_blog (TODO: Find t3blog "parent_id" comment, find equivalent in mm_forum and quote the post)
- Posts edited in mm_forum are not edited in t3blog comments (TODO: use mm_forum hooks)
- New post in the mm_forum topic does not create a new comment in t3blog (TODO: use mm_forum hooks) => Make topics or (better) the whole forum (you choose to let the comments be created in) read only
- You may get two E-Mail notifications: One by mm_forum and one by t3blog. 
- editing comments is not supported, neither in FE or BE

ChangeLog
2010-05-30 Hauke Hain  <hhpreuss@googlemail.com>

	* Added possibility to use the t3blog commenting interface. By default 
	  mm_forum_blog creates a mm_forum topic if the first comments is written.
	  If comments are immediately approved a post is created.
	  
	  Attention: The BE module "Comments" of t3blog is not supported at all!
	
2010-05-24 Hauke Hain  <hhpreuss@googlemail.com>

	* first working version: replaces the t3blog commenting widget with mm_forum_comments	

ext_localconf.php
<?php

	/*
	 * Add some additional page TSConfig. This is used to dynamically extend the
	 * mm_forum backend module.
	 */
t3lib_extMgm::addPageTSConfig('<INCLUDE_TYPOSCRIPT: source="FILE:EXT:mm_forum_blog/res/ts/tx_mmforumblog_pagetsconfig.ts">');

//mm_forum_comments
$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['mm_forum_comments']['getTypoScriptDataHook'][$_EXTKEY] = 'EXT:mm_forum_blog/lib/class.tx_mmforumcomments_hooks_for_blog.php:&tx_mmforumcomments_hooks_for_blog->getTypoScriptDataHook';
$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['mm_forum_comments']['overwriteParameter'][$_EXTKEY] = 'EXT:mm_forum_blog/lib/class.tx_mmforumcomments_hooks_for_blog.php:&tx_mmforumcomments_hooks_for_blog->overwriteParameter';

//t3blog
$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['t3blog']['aftercommentinsertion'][$_EXTKEY] = 'EXT:mm_forum_blog/lib/class.tx_mmforumcomments_hooks_for_blog.php:&tx_mmforumcomments_hooks_for_blog->t3blog_aftercommentinsertion';

//mm_forum (author name for t3blog)
$TYPO3_CONF_VARS['EXTCONF']['mm_forum']['display']['postListMarkerArray'][] = 'EXT:mm_forum_blog/lib/class.tx_mmforumcomments_hooks_for_blog.php:&tx_mmforumcomments_hooks_for_blog';

?>

lib/class.tx_mmforumblog_div.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 Hauke Hain <hhpreuss@googlemail.com>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

class tx_mmforumblog_div {

  public static function html2bbcode($html) {
    $html = str_replace(chr(10), '', $html); //remove carriage returns
    $html = str_replace(chr(13), '', $html); //remove carriage returns
    return tx_mmforumblog_div::br2nl($html);
  }

  public static function br2nl($html) {
    return preg_replace('#<br\s*/?>#i', "\n", $html);
  }

}

?>

res/ts/tx_mmforumblog_pagetsconfig.ts
mod.web_txmmforumM1 {
	essentialConfiguration {
		mmforumblog_blogCommentCategory = 1
		mmforumblog_blogTopicAuthor     = 1
	}

	submodules {
		installation {
			categories {
				tt_blog = MMFORUM_CONF_CATEGORY
				tt_blog {
					icon  = EXT:mm_forum_blog/res/img/mod/mmforum-conf.png
					name  = LLL:EXT:mm_forum_blog/res/lang/locallang.xml:mod.category_title_short
					title = LLL:EXT:mm_forum_blog/res/lang/locallang.xml:mod.category_title

					items {
						mmforumblog_blogCommentCategory = MMFORUM_CONF_ITEM
						mmforumblog_blogCommentCategory {
							type = special
							type.handler = EXT:mm_forum_blog/lib/class.tx_mmforumblog_modinstall.php:tx_mmforumblog_ModInstall->getForumSelector

							label       = LLL:EXT:mm_forum_blog/res/lang/locallang.xml:mod.blogCommentCategory.title
							description = LLL:EXT:mm_forum_blog/res/lang/locallang.xml:mod.blogCommentCategory.desc
						}

						mmforumblog_blogTopicAuthor = MMFORUM_CONF_ITEM
						mmforumblog_blogTopicAuthor {
							type = group
							type.table = fe_users

							label       = LLL:EXT:mm_forum_blog/res/lang/locallang.xml:mod.blogTopicAuthor.title
							description = LLL:EXT:mm_forum_blog/res/lang/locallang.xml:mod.blogTopicAuthor.desc
						}
					}
				}
			}
		}
	}

}

lib/class.tx_mmforumcomments_hooks.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 Hauke Hain <hhpreuss@googlemail.com>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

#include_once(t3lib_extMgm::extPath('t3blog').'pi1/lib/class.t3blog_div.php');

class tx_mmforumcomments_hooks {
  function mainDataHook(&$data, &$pObj) {
    // get all content elemenets
			$resContent = $GLOBALS['TYPO3_DB']->exec_SELECTquery(
				'uid, bodytext',		// SELECT ...
				'tt_content',		// FROM ...
				'irre_parentid = ' . $data['uid'] .
					' AND irre_parenttable=\'tx_t3blog_post\' ' .
					$pObj->cObj->enableFields('tt_content'),
				'uid',		// GROUP BY ...
				'sorting'		// ORDER BY ...
			);

			$uidContentList = '';
			$divider = false;
			while (false !== ($rowContent = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($resContent)) && !$divider) {
				// seperate by divider
				$dividerInText 		= stripos($rowContent['bodytext'], '###MORE###');
				$textBeforeDivider= substr($rowContent['bodytext'], 0, strpos($rowContent['bodytext'], '###MORE###'));

				if($dividerInText > 0 ) {
					$divider = true;
				}	else {
					$uidContentList .= intval($rowContent['uid']).',';
				}
			}
			$GLOBALS['TYPO3_DB']->sql_free_result($resContent);

			if (strlen($uidContentList) > 1) {
				$uidContentList = substr($uidContentList,0,strlen($uidContentList)-1);
			}

    // save data
    $data['blogPid'] = t3blog_div::getBlogPid();
    $data['contentUids'] = $uidContentList;
    $data['textBeforeDivider'] = strip_tags($textBeforeDivider);
  }
}

?>

lib/class.tx_mmforumblog_modinstall.php
<?php

/*                                                                     *
 *  COPYRIGHT NOTICE                                                   *
 *                                                                     *
 *  (c) 2010 Martin Helmich <m.helmich@mittwald.de>                    *
 *      All rights reserved                                            *
 *                                                                     *
 *  This script is part of the TYPO3 project. The TYPO3 project is     *
 *  free software; you can redistribute it and/or modify               *
 *  it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation; either version 2 of the License,  *
 *  or (at your option) any later version.                             *
 *                                                                     *
 *  The GNU General Public License can be found at                     *
 *  http://www.gnu.org/copyleft/gpl.html.                              *
 *                                                                     *
 *  This script is distributed in the hope that it will be useful,     *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of     *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the      *
 *  GNU General Public License for more details.                       *
 *                                                                     *
 *  This copyright notice MUST APPEAR in all copies of the script!     *
 *                                                                     */



	/**
	 *
	 * This class provides functions for extending the mm_forum backend module.
	 *
	 * @author    Martin Helmich <m.helmich@mittwald.de>
	 * @copyright 2009 Martin Helmich <m.helmich@mittwald.de>
	 * @version   2009-12-22
	 * @package   mm_forum_news
	 *
	 */

class tx_mmforumblog_ModInstall {



		/**
		 *
		 * Generates a custom board selection field.
		 * This method generates a <SELECT> element that presents all message
		 * boards in a hierarchical order. This input field will be presented in
		 * the mm_forum configuration backend module.
		 *
		 * @author  Martin Helmich <m.helmich@mittwald.de>
		 * @version 2009-12-22
		 * @param   array  $value     The current value of this input field. The
		 *                            option with this value will be selected.
		 * @param   string $fieldname The name of the input field.
		 * @param   array  $config    A configuration array with additional
		 *                            properties for this field. In this case,
		 *                            this variable is not used.
		 * @return  string            The <SELECT> input field as HTML output.
		 *
		 */

	public function getForumSelector($value, $fieldname, $config) {
		global $TYPO3_DB;
		$content = '';
		$pid = $this->p->conf['storagePID'];

			/*
			 * Get all categories. Iterate over the categories and load the
			 * boards for each category.
			 */
		$res = $TYPO3_DB->exec_SELECTquery( 'uid, forum_name',
		                                    'tx_mmforum_forums',
			                                'parentID = 0 AND deleted=0 AND pid='.$pid, '', 'sorting ASC');
		while($arr = $TYPO3_DB->sql_fetch_assoc($res)) {
			$res2 = $TYPO3_DB->exec_SELECTquery( 'uid, forum_name',
                                                 'tx_mmforum_forums',
			                                     "parentID = {$arr[uid]} AND deleted=0 AND pid=$pid", '', 'sorting ASC' );
			if($TYPO3_DB->sql_num_rows($res2)>0) {
				$content .= '<optgroup label="'.htmlspecialchars($arr['forum_name']).'">'.chr(10);
				while($arr2 = $TYPO3_DB->sql_fetch_assoc($res2))
					$content .= '<option value="'.$arr2['uid'].'" '.($value==$arr2['uid']?'selected="selected"':'').'>'
					            . htmlspecialchars($arr2['forum_name'])
								. '</option>'.chr(10);
				$content .= '<optgroup>'.chr(10);
			}
		}

		return '<select name="tx_mmforum_install[conf][0]['.htmlspecialchars($fieldname).']">'.$content.'</select>';
	}

}

?>

lib/class.tx_mmforumcomments_hooks_for_blog.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 Hauke Hain <hhpreuss@googlemail.com>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

include_once(t3lib_extMgm::extPath('t3blog').'pi1/lib/class.t3blog_div.php');
include_once(t3lib_extMgm::extPath('mm_forum_comments').'lib/class.tx_mmforumcomments_div.php');

class tx_mmforumcomments_hooks_for_blog {
  /**
 * Manipulates the result of "getTypoScriptData" - adding t3blog fields
 *
 * @param	array		$parameter: contains the data for TypoScript, recordsTable, key, uid, TS setup
 * @param	array		$pObj: parent -> calling object ($this)
 * @return	null		manipulates the original data because of parameters by reference
 */
  function getTypoScriptDataHook(&$parameter, &$pObj) {
    if(intval($parameter['data']['uid'])==0) {
      return;
    }

    if($pObj==false) {
      $where = '';
    } else {
      $where = $pObj->cObj->enableFields('tt_content');
    }

    // get all content elemenets
  	$resContent = $GLOBALS['TYPO3_DB']->exec_SELECTquery(
  				'uid, bodytext',		// SELECT ...
  				'tt_content',		// FROM ...
  				'irre_parentid = ' . intval($parameter['data']['uid']) .
  					' AND irre_parenttable=\'tx_t3blog_post\' ' . $where,
  				'uid',		// GROUP BY ...
  				'sorting'		// ORDER BY ...
  	);

  	$uidContentList = '';
  	$textBeforeDivider = '';
  	$divider = false;

  	while (false !== ($rowContent = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($resContent)) && !$divider) {
  		// seperate by divider
  		$dividerInText 		= stripos($rowContent['bodytext'], '###MORE###');

  		if($dividerInText===false) {
  			$textBeforeDivider .= $rowContent['bodytext'].' ';
      } else {
        $textBeforeDivider .= substr($rowContent['bodytext'], 0, strpos($rowContent['bodytext'], '###MORE###')).' ';
      }

  		if($dividerInText > 0 ) {
  			$divider = true;
  		}	else {
  			$uidContentList .= intval($rowContent['uid']).',';
  		}
  	}

    $GLOBALS['TYPO3_DB']->sql_free_result($resContent);

  	if (strlen($uidContentList) > 1) {
  		$uidContentList = substr(nl2br($uidContentList,0,strlen($uidContentList)-1));
  	}

    // save data
    $parameter['data']['blogPid'] = t3blog_div::getBlogPid();
    $parameter['data']['contentUids'] = $uidContentList;
    $parameter['data']['textBeforeDivider'] = strip_tags($textBeforeDivider);
  }

	/**
	 * "getParameter" returns the result of this function
	 *
	 * @param	array		$paraconf: The parameter segment of the TS plugin setup
	 * @param	array		$pObj: parent -> calling object ($this)
	 * @return	array		database uid given by postVar and the parameter key to look in TypoScript parameter section
	 */
  function overwriteParameter(&$paraconf, &$pObj) {
    foreach($paraconf as $key => $value) {
      $key = substr($key, 0, strlen($key)-1);
      $uidParent = $value['uidParent'];
      $uidkey = $value['uid'];
      $gp = t3lib_div::_GP($key);

      if (!empty($gp[$uidkey])) {
        return array($key . '->' . $uidkey, $gp[$uidkey], $key);
      } elseif (!empty($gp[$uidParent][$uidkey])) {
        return array($key . '->' . $uidParent . '->' . $uidkey, $gp[$uidParent][$uidkey], $key);
      }
    }
    return array();
  }

	/**
	 * Creates a new commenttopic if none exists. Create a new post if comment
	 * is approved.	 
	 *
	 * @param	array  		$$paraconf: ...
	 * @param	object		$pObj: parent object
	 * @return	null		returns nothing
	 */
  function t3blog_aftercommentinsertion(&$paraconf, &$pObj) {
    $setup = tx_mmforumcomments_div::loadTSSetupForPage($GLOBALS['TSFE']->id);
    $parameters = array('tx_t3blog_pi1->blogList->showUid',$paraconf['postUid'],'tx_t3blog_pi1');
    $topicID = tx_mmforumcomments_div::getTopicID($GLOBALS['TSFE']->id, $parameters);

    //create new mm_forum topic if none exists
    if ($topicID === 0) {
      tx_mmforumcomments_div::createTopicForRecord($parameters, $setup['plugin.']['tx_mmforumcomments_pi1.'], $GLOBALS['TSFE']->id, $setup['plugin.']['tx_mmforum.']['storagePID'], $pObj);
    }

    if (($pObj->conf['approved']) && !($topicID === 0)){
      //create new comment in mm_forum
      include_once(t3lib_extMgm::extPath('mm_forum_comments').'lib/class.tx_mmforumcomments_createcomments.php');
      include_once(t3lib_extMgm::extPath('mm_forum_blog').'lib/class.tx_mmforumblog_div.php');
      $comAuthor = (intval($GLOBALS['TSFE']->fe_user->user['uid']) == 0) ? tx_mmforumcomments_div::getTopicAuthorUID($parameters[2], $setup['plugin.']['tx_mmforumcomments_pi1.']) : intval($GLOBALS['TSFE']->fe_user->user['uid']);
      $text = tx_mmforumblog_div::html2bbcode($pObj->cObj->stdWrap($paraconf['data']['title'], $setup['plugin.']['tx_mmforumcomments_pi1.']['t3blogTitleInCommentWrap.'])) . "\n" . tx_mmforumblog_div::html2bbcode($paraconf['data']['text']);
      tx_mmforumcomments_createcomments::createPost($topicID, $comAuthor,
                            $text,
                            $paraconf['data']['tstamp'],
                            $setup['plugin.']['tx_mmforum.']['storagePID'],
                            $paraconf['data']['author']
                        );
    }
  }

	/**
	 * Replaces the Username with the one of t3blog. Changes the postmenu so that
	 * the post looks like from a deleted user.    	 
	 *
	 * @param	array  		$marker: marker array
	 * @param	array  		$row: post data
	 * @param	array  		$topic: topic data
	 * @param	object		$obj: calling object
	 * @return	array		$marker
	 */
	public function processPostListMarkerArray($marker, $row, $topic, $obj){
    if (strlen(trim($row['tx_mmforumcomments_authorname']))>0) {
      $marker['###POSTUSER###'] = $obj->cObj->stdWrap($row['tx_mmforumcomments_authorname'], $obj->conf['list_posts.']['userinfo.']['username_stdWrap.']);

      //assuming that $topic['topic_poster'] is the same as tx_mmforumcomments_div::getTopicAuthorUID
      if($topic['topic_poster']===$row['poster_id']) {
        $marker['###POSTMENU###'] = $marker['###MESSAGEMENU###'];
        $marker['###POSTUSER###'] .= $obj->pi_getLL('user-deleted');
      }
    }
		return $marker;
	}

}

?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "mm_forum_blog".
#
# Auto generated 30-05-2010 19:09
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Blog comments with mm_forum',
	'description' => 't3blog comments are also written to the mm_forum topic created by mm_forum_comments.',
	'category' => 'fe',
	'author' => 'Hauke Hain',
	'author_email' => 'hhpreuss@googlemail.com',
	'shy' => '',
	'dependencies' => 'cms,mm_forum,mm_forum_comments,t3blog',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'alpha',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.0.0',
	'constraints' => array(
		'depends' => array(
			'cms' => '',
			'mm_forum' => '1.9.0-0.0.0',
			'mm_forum_comments' => '1.1.0-0.0.0',
			't3blog' => '0.9.1-0.0.0',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:16:{s:9:"ChangeLog";s:4:"c62f";s:10:"README.txt";s:4:"445c";s:12:"ext_icon.gif";s:4:"167d";s:17:"ext_localconf.php";s:4:"0ae6";s:14:"ext_tables.php";s:4:"632d";s:14:"ext_tables.sql";s:4:"26ee";s:19:"doc/wizard_form.dat";s:4:"227e";s:20:"doc/wizard_form.html";s:4:"0a46";s:32:"lib/class.tx_mmforumblog_div.php";s:4:"d5cb";s:39:"lib/class.tx_mmforumblog_modinstall.php";s:4:"eb3b";s:47:"lib/class.tx_mmforumcomments_hooks_for_blog.php";s:4:"d7dc";s:28:"res/img/mod/mmforum-conf.png";s:4:"de58";s:22:"res/lang/locallang.xml";s:4:"61ab";s:37:"res/ts/tx_mmforumblog_pagetsconfig.ts";s:4:"9577";s:23:"res/ts/static/setup.txt";s:4:"6c18";s:30:"res/ts/staticReplace/setup.txt";s:4:"78c3";}',
	'suggests' => array(
	),
);

?>